src/com/rcv/Main.java

```diff
@@ -164,31 +163,84 @@ static String executeTabulation(ElectionConfig config) {
+  // returns: list of all CastVoteRecord objects parsed from CVR files (or null if there's an error)
+  private static List<CastVoteRecord> parseCastVoteRecords(ElectionConfig config) {
+    // castVoteRecords will contain all cast vote records parsed by the reader
+    List<CastVoteRecord> castVoteRecords = new ArrayList<>();
+    // how many sources had fatal problems?
+    int numInvalidSources = 0;
+
     if (config.rawConfig.cvrFileSources == null || config.rawConfig.cvrFileSources.isEmpty()) {
       Logger.severe("Config doesn't contain any CVR input files.");
+    } else {
+      // At each iteration of the following loop, we add records from another source file.
+      // source: index over config sources
+      for (RawElectionConfig.CVRSource source : config.rawConfig.cvrFileSources) {
+        Logger.info("Reading CVR file: %s (provider: %s)", source.filePath, source.provider);
+
+        // did we encounter a fatal problem for this source?
+        boolean encounteredProblemForThisSource = false;
+
+        if (source.filePath == null || source.filePath.isEmpty()) {
+          Logger.severe(
+              "Invalid source file: missing filePath"
+          );
+          encounteredProblemForThisSource = true;
+        }
+
+        if (source.firstVoteColumnIndex == null || source.firstVoteColumnIndex < 0) {
+          Logger.severe(
+              "Invalid source file: missing or invalid firstVoteColumnIndex: %s",
+              source.filePath
+          );
+          encounteredProblemForThisSource = true;
+        }
+
+        if (
+            config.isTabulateByPrecinctEnabled() &&
+            (source.precinctColumnIndex == null || source.precinctColumnIndex < 0)
+        ) {
+          Logger.severe(
+              "Invalid source file: missing or invalid precinctColumnIndex when " +
+                  "tabulateByPrecinct is enabled: %s",
+              source.filePath
+          );
+          encounteredProblemForThisSource = true;
+        }
+
+        if (!encounteredProblemForThisSource) {
+          // reader: read input file into a list of cast vote records
+          CVRReader reader = new CVRReader(
+              config,
+              source.filePath,
+              source.firstVoteColumnIndex,
+              source.idColumnIndex,
+              source.precinctColumnIndex
+          );
+          // the CVRs parsed from this source
+          List<CastVoteRecord> cvrs = reader.parseCVRFile();
+          if (cvrs.isEmpty()) {
+            Logger.severe("Source file contains no CVRS: %s", source.filePath);
+            encounteredProblemForThisSource = true;
+          }
+          // add records to the master list
+          castVoteRecords.addAll(cvrs);
+        }
+
+        if (encounteredProblemForThisSource) {
+          numInvalidSources++;
+        }
+      }
+      if (numInvalidSources > 0) {
+        Logger.severe("Encountered %d invalid source(s)", numInvalidSources);
```

Comment:
Should we add something here about the process terminating entirely as a result of invalid sources?

Reply:
When this happens, the next log line will say, "Skipping tabulation due to source file errors."